File: Backend/Schemas/taller.py
from pydantic import BaseModel
from datetime import datetime
from typing import List, Optional
from ..Models.optica import EstadoOrden

# --- Schemas para Etapas de Trabajo ---

class EtapaTrabajoBase(BaseModel):
    etapa: str
    completado: bool = False
    tecnico_id: Optional[int] = None
    notas: Optional[str] = None

class EtapaTrabajoCreate(EtapaTrabajoBase):
    pass

class EtapaTrabajoResponse(EtapaTrabajoBase):
    id: int
    orden_id: int
    fecha_creacion: datetime
    fecha_actualizacion: datetime

    class Config:
        from_attributes = True

# --- Schemas para Histórico de Estados ---

class HistoricoEstadosResponse(BaseModel):
    id: int
    estado_anterior: Optional[EstadoOrden] = None
    estado_nuevo: EstadoOrden
    tecnico_id: Optional[int] = None
    fecha_creacion: datetime

    class Config:
        from_attributes = True

# --- Schemas para Órdenes de Trabajo ---

class OrdenTrabajoBase(BaseModel):
    descripcion_trabajo: Optional[str] = None
    fecha_entrega_esperada: Optional[datetime] = None

class OrdenTrabajoCreate(OrdenTrabajoBase):
    venta_id: int

class OrdenTrabajoResponse(OrdenTrabajoBase):
    id: int
    venta_id: int
    estado: EstadoOrden
    etapas: List[EtapaTrabajoResponse] = []
    historico_estados: List[HistoricoEstadosResponse] = []

    class Config:
        from_attributes = True

class CambiarEstadoOrden(BaseModel):
    estado_nuevo: str
    tecnico_id: Optional[int] = None

# --- Schema para el endpoint de demo del Composite Pattern ---

class EtapaOrdenData(BaseModel):
    nombre: str
    tiempo_estimado: int

File: visualion-frontend/types/taller.ts
// types/taller.ts
export type EstadoOrden = "recibida" | "en_proceso" | "lista_para_entrega" | "entregada" | "cancelada";

export interface OrdenTrabajo {
  id: number;
  venta_id: number;
  estado: EstadoOrden;
  descripcion_trabajo?: string;
  fecha_entrega_esperada?: string;
  fecha_creacion: string;
  fecha_actualizacion: string;
  etapas?: EtapaTrabajo[];
  historico_estados?: HistoricoEstados[];
}

export interface EtapaTrabajo {
  id: number;
  orden_id: number;
  etapa: string;
  tecnico_id?: number;
  completado: boolean;
  notas?: string;
  fecha_creacion?: string;
  fecha_actualizacion?: string;
}

export interface HistoricoEstados {
  id: number;
  orden_id: number;
  estado_anterior?: string;
  estado_nuevo: string;
  tecnico_id?: number;
  fecha_creacion: string;
}

export const ESTADOS_COLORES: Record<EstadoOrden, string> = {
  recibida: "bg-gray-600",
  en_proceso: "bg-yellow-600",
  lista_para_entrega: "bg-blue-600",
  entregada: "bg-green-600",
  cancelada: "bg-red-600"
};

export const ESTADOS_LABEL: Record<EstadoOrden, string> = {
  recibida: "Recibida",
  en_proceso: "En Proceso",
  lista_para_entrega: "Lista para Entrega",
  entregada: "Entregada",
  cancelada: "Cancelada"
};
